### PR TITLE
feat: implement api breaker plugin.

### PR DESCRIPTION
**Pull request by membphis**
### What this PR does / why we need it:

Users can configure healthy and unhealthy states and the number of occurrences.

When the unhealthy state is triggered, apisix considers the upstream service to be in an unserviceable state and returns an error directly to the client, and does not request the upstream service again.

The duration of the unserviceable state increases exponentially with the number of times it is triggered.

When the number of healthy state arrivals is triggered, the normal state of the API is restored.

continue the job: https://github.com/apache/apisix/pull/2339

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible?

### CODE REVIEW

File: apisix/plugins/api-breaker.lua [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                }
+            },
+            default = {http_statuses = {200}, successes = 3}
+        }
+    },
+    required = {"break_response_code"},
+}
+
+
+-- todo: we can move this into `core.talbe`
```
Comment (moonming) [inline apisix/plugins/api-breaker.lua:93]: please create a issue for this, only comment is not enough

File: apisix/plugins/api-breaker.lua [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    -- cannot exceed the maximum value of the user configuration
+    local breaker_time = 2 ^ failure_times
+    if breaker_time > conf.max_breaker_sec then
+        breaker_time = conf.max_breaker_sec
+    end
+    core.log.info("breaker_time: ", breaker_time)
+
+    -- breaker
+    if unhealthy_lastime + breaker_time >= ngx.time() then
```
Comment (moonming) [inline apisix/plugins/api-breaker.lua:173]: what is `lastime`?

File: t/plugin/api-breaker.t [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+GET /t
+--- response_body
+{"break_response_code":502,"healthy":{"http_statuses":[200],"successes":3},"max_breaker_sec":300,"unhealthy":{"failures":3,"http_statuses":[500]}}
+--- no_error_log
+[error]
+--- LAST
+
+
+
+=== TEST 5: add plugin
```
Comment (moonming) [inline t/plugin/api-breaker.t:141]: move this test case before test 10

File: t/plugin/api-breaker.t [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+            ngx.say(require("lib.json_sort").encode(conf))
+        }
+    }
+--- request
+GET /t
+--- response_body
+{"break_response_code":502,"healthy":{"http_statuses":[200],"successes":3},"max_breaker_sec":300,"unhealthy":{"failures":3,"http_statuses":[500]}}
+--- no_error_log
+[error]
+--- LAST
```
Comment (moonming) [inline t/plugin/api-breaker.t:137]: remove it

**Review by moonming — CHANGES_REQUESTED** [review]

File: apisix/plugins/api-breaker.lua (hunk 1 above)
Comment (membphis) [reply apisix/plugins/api-breaker.lua:93]: related issue: https://github.com/apache/apisix/issues/2486

**Comment by membphis** [comment]
@moonming I have fixed all of the points, you can take a look when you have time

**Review by moonming — APPROVED** [review]

**Review by moonming — COMMENTED** [review]
please add doc in README and plugin list.

**Review by gxthrj — APPROVED** [review]

**Comment by membphis** [comment]
> please add doc in README and plugin list.

added a todo issue: https://github.com/apache/apisix/issues/2529

**Pull request merged**